src/main/java/org/snomed/snowstorm/core/data/services/ConceptChangeHelper.java

```diff
@@ -255,128 +255,153 @@ private NativeSearchQueryBuilder newSearchQuery(BoolQueryBuilder branchUpdatesCr
-        // has in fact been deleted on the task.
-        Set<Long> conceptsNotReplaced = new HashSet<>();
-        Set<String> versionsReplaced = branchService.findLatest(branchPath).getVersionsReplaced(Concept.class);
-        if (!versionsReplaced.isEmpty()) {
-            Set<Long> conceptIds = getConceptsByInternalIdentifierAndPath(versionsReplaced);
+        // Case 2
+        if (!conceptsReplacedOnBranch.isEmpty()) {
             try (final SearchHitsIterator<Concept> stream = elasticsearchTemplate.searchForStream(new NativeSearchQueryBuilder()
                     .withQuery(
                             boolQuery()
-                                    .must(termsQuery(Concept.Fields.CONCEPT_ID, conceptIds))
+                                    .must(termsQuery(Concept.Fields.CONCEPT_ID, conceptsReplacedOnBranch))
                                     .must(termQuery(Concept.Fields.PATH, branchPath))
                                     .mustNot(existsQuery(Concept.Fields.END))
                     )
+                    .withFields(Concept.Fields.CONCEPT_ID)
                     .build(), Concept.class)) {
-                stream.forEachRemaining(hit -> {
-                    String id = hit.getContent().getConceptId();
-                    // Concept replacement found, therefore not deleted.
-                    if (id != null) {
-                        conceptIds.removeIf(e -> e.equals(Long.parseLong(id)));
-                    }
-                });
+                stream.forEachRemaining(hit -> conceptsReplacedOnBranch.removeIf(id -> id.equals(hit.getContent().getConceptId())));
             }
+        }
+
+        conceptsDeleted.addAll(conceptsReplacedOnBranch);
+
+        return conceptsDeleted;
+    }
 
-            conceptsNotReplaced.addAll(conceptIds);
+    private List<String> getAncestors(String branchPath) {
+        if (branchPath == null || branchPath.isEmpty() || isCodeSystemBranch(branchPath)) {
+            return Collections.emptyList();
         }
 
-        conceptsDeletedOnBranch.addAll(conceptsNotReplaced);
+        List<String> branchPathHierarchy = getBranchPathHierarchy(branchPath);
+        int hierarchySize = branchPathHierarchy.size();
+        List<String> parts = new ArrayList<>();
+        boolean isCodeSystemBranch;
+        for (int x = 0; x < hierarchySize; x++) {
+            String part = branchPathHierarchy.get(x);
+            isCodeSystemBranch = isCodeSystemBranch(part);
 
-        return conceptsDeletedOnBranch;
+            if (!isLastIteration(x, hierarchySize)) {
+                part = prependParentParts(x, hierarchySize, branchPathHierarchy, part);
+            }
+
+            if (!Objects.equals(part, branchPath)) {
+                parts.add(part);
+            }
+
+            // Do not process any more ancestors, e.g.
+            // MAIN or MAIN/SNOMEDCT-XX will be the root for the given branchPath.
+            if (isCodeSystemBranch) {
+                break;
```

Comment:
We could make this simpler.

Reply:
We can use PathUtil.getParentPath() to get a list of ancestors instead doing it from scratch.

Reply:
@CoderMChu, that will only return the immediate parent and not all of the ancestors. Also, it'll return `main` when asking for the parent of `main/snomedct-xx`. There shouldn't be any ancestors/parent of `main/snomedct-xx`.

Reply:
You need to stop processing when current branch is a code system branch.(i.e stop when main or main/snomedct-xx)

Reply:
Here is an example:
`private List<String> getAncestors(String branch) {
   List<String> ancestors = new ArrayList<>();
   String parent = PathUtil.getParentPath(branch);
   ancestors.add(parent);
   while (!isCodeSystemBranch(parent)) {
      parent = PathUtil.getParentPath(parent);
      ancestors.add(parent);
   }
   return ancestors;
}`